sherpa/utils/__init__.py

```diff
@@ -1327,71 +1327,209 @@ def print_fields(names, vals, converters=None):
-                expr.append(format % vals[ii])
-            if expr[-1] != delim:
-                expr.append(delim)
-        else:
-            if not expr[-1] in (',', delim):
-                expr.append(',')
-            expr.append(format % vals[ii])
-            expr.append(',')
-    if len(expr) and expr[-1] in (',', delim):
-        expr.append(format % vals[-1])
 
-    return ''.join(expr)
+        # If we have a break then output the range we have just
+        # processed.
+        #
+        expr.append(filt(start, end))
+
+        # Start a new range.
+        #
+        start = val
+        end = val
+
+    # Handle the last range.
+    #
+    expr.append(filt(start, end))
+
+    return ','.join(expr)
 
 
 def parse_expr(expr):
+    """Convert a filter expression into its parts.
+
+    This is intended for parsing a notice or ignore expression
+    given as a string.
+
+    Parameters
+    ----------
+    expr : str
+        The filter expression, of the form 'a:b' or a single number,
+        separated by commas, and white space is ignored. The
+        upper or lower limit of a pair may be ignored (e.g. 'a:' or
+        ':b').
+
+    Returns
+    -------
+    filters : list of pairs
+        Each pair gives the lower- and upper-edge of the filter,
+        using ``None`` to represent no limit.
+
+    Notes
+    -----
+    There is no attempt to validate that the expression contains
+    strictly ordered pairs, or that the pairs do not overlap, or
+    that the lower- and upper-limits are in increasing numerical
+    order. That is, the expression '5:7,:2,4:6,5:3' is allowed.
+
+    Examples
+    --------
+
+    >>> parse_expr('0.5:7')
+    [(0.5, 7.0)]
+
+    >>> parse_expr('0.5:')
+    [(0.5, None)]
+
+    >>> parse_expr(':7')
+    [(None, 7.0)]
+
+    >>> parse_expr(':2, 4 : 5 ,7:8,10:')
+    [(None, 2.0), (4.0, 5.0), (7.0, 8.0), (10.0, None)]
+
+    >>> parse_expr('4')
+    [(4.0, 4.0)]
+
+    >>> parse_expr(' ')
+    [(None, None)]
+
     """
-    parse a filter expression into numerical components for notice/ignore
-    e.g. ':2,4:5,7:8,10:'
-    """
-    res = []
+
     if expr is None or str(expr).strip() == '':
-        res.append((None, None))
-        return res
+        return [(None, None)]
+
+    res = []
     vals = str(expr).strip().split(',')
     for val in vals:
         lo, hi = None, None
+
         interval = val.strip().split(':')
-        if len(interval) == 1:
+        ninterval = len(interval)
+        if ninterval == 1:
             lo = interval[0]
             if lo == '':
                 lo = None
             hi = lo
-        elif len(interval) > 1:
+        elif ninterval == 2:
             lo = interval[0]
             hi = interval[1]
             if lo == '':
                 lo = None
             if hi == '':
                 hi = None
         else:
+            # This check exited but was never hit due to the way the
+            # code was written. It now errors out if a used gives
```

Comment:
'used' or 'user'

Reply:
It should be 'user'.

Reply:
I've just pushed a commit to fix the typo. This comment is probably too defensive but I just wanted to point out that there is a subtle change in behavior here. I note that the "user-level" documentation - e.g. sherpa.astro.ui.notice - only ever mentions "a", "a:b", "a:", or ":b" for these fragments (so we shouldn't expect anything more).